File: src/ios/Store.swift


class Store {
    private var store:[String:Any]=[:];
    
    init(){
        
    }
    
    func initiate(_ state:[String:Any]?){
        if(state != nil){
            self.store=state!;
        }
    }
    
    func get(_ path:String)throws->[Any]{
        let keys=path.replacingOccurrences(of:" ",with:"").components(separatedBy:[".","["]);
        var data:[Any]=[store];
        try keys.forEach({key in
            data=try data.flatMap({item in try Store.getProperties(key,item)});
        });
        try self.mutate(Store.getKeysPath(keys),data);
        return data;
    }
    
    func set(_ path:String,_ value:Any?=nil) throws {
        if(path.isEmpty){
            throw Webview.Error("invalid key");
        }
        else{
            try self.mutate(path,[value ?? false]);
        }
    }
    
    func delete(_ path:String) throws {
        var keys=path.replacingOccurrences(of:" ",with:"").components(separatedBy:[".","["]);
        let target=keys.popLast()!;
        var data:[Any]=[store];
        try keys.forEach({key in
            data=try data.flatMap({item in try Store.getProperties(key,item)});
        });
        
        let length=data.count;
        for i in 0..<length {
            try Store.deleteProperty(target,&data[i]);
        }
        try self.mutate(Store.getKeysPath(keys),data);
    }
    
    private func mutate(_ path:String,_ values:[Any]=[false]) throws {
        var keys=path.replacingOccurrences(of:" ",with:"").components(separatedBy:[".","["]);
        let target=keys.popLast()!;
        var data:[Any]=[store];
        try keys.forEach({key in
            data=try data.flatMap({item in try Store.getProperties(key,item)});
        });
        
        let length=data.count;
        for i in 0..<length {
            try Store.setProperty(target,target.contains("*") ? values:values[0],&data[i]);
        }
        if(keys.isEmpty){
            store=data[0] as? [String:Any] ?? [:];
        }
        else{
            try self.mutate(Store.getKeysPath(keys),data);
        }
    }
    
    static func getKeysPath(_ keys:[String])->String{
        var path="";
        keys.forEach({key in
            path+=(key.hasSuffix("]") ? "[":".")+key;
        });
        path=String(path[path.index(after:path.startIndex)...]);
        return path;
    }
    
    static func getProperties(_ key:String,_ data:Any)throws->[Any]{
        if let value=data as? Bool,value==false {
            throw Webview.Error("cannot get properties of falsy (getting \""+key+"\")");
        }
        var props:[Any]=[];
        let index=key.firstIndex(of:"]");
        if let bracketIndex=index {
            if var array=data as? [Any?] {
                let symbol=key.prefix(upTo:bracketIndex);
                switch(symbol){
                    case "*":
                        array.forEach({item in
                            props.append(item ?? false);
                        });
                        break;
                    case "pop":
                        props.append(array.removeLast() ?? false);
                        break;
                    case "shift":
                        if(array.isEmpty){
                            props.append(false);
                        }
                        else{
                            props.append(array.removeFirst() ?? false);
                        }
                        break;
                    case "last":
                        let index=array.count-1;
                        if(index > -1){
                            props.append(array[index] ?? false);
                        }
                        break;
                    default:
                        let index=Int(symbol) ?? -1;
                        if((0..<array.count).contains(index)){
                            props.append(array[index] ?? false);
                        }
                        else{
                            props.append(false);
                        }
                        break;
                }
            }
            else{
                throw Webview.Error("object cannot be cast to array");
            }
        }
        else if let dictionary=data as? [String:Any] {
            props.append(dictionary[key] ?? false);
        }
        else{
            throw ObjectCastError;
        }
        return props;
    }
    
    static func setProperty(_ key:String,_ value:Any?,_ data: inout Any) throws {
        if let value=data as? Bool,value==false {
            throw Webview.Error("cannot set properties of falsy (setting \""+key+"\")");
        }
        if let bracketIndex=key.firstIndex(of:"]"){
            if var array=data as? [Any?] {
                let symbol=key.prefix(upTo:bracketIndex);
                switch(symbol){
                    case "push":array.append(value ?? false);break;
                    case "unshift":
                        array.insert(value ?? false,at:0);break;
                    case "pop":
                        _=array.popLast();break;
                    case "shift":
                        if(!array.isEmpty){
                            array.remove(at:0);
                        }
                        break;
                    case "last":
                        array[array.count-1]=value ?? false;break;
                    case "*":
                        let length=array.count;
                        let values=value as! [Any];
                        for i in 0..<length {
                            array[i]=values[values.count>1 ?i:0];
                        }
                        break;
                    default:
                        let index=Int(symbol)!;
                        array[index]=value ?? false;
                        break;
                }
                data=array;
            }
            else{
                throw Webview.Error("Object caanot be cast to array");
            }
        }
        else if var dictionary=data as? [String:Any]{
            dictionary[key]=value;
            data=dictionary;
        }
        else{
            throw ObjectCastError;
        }
    }
    
    static func deleteProperty(_ key:String,_ data: inout Any) throws {
        if let value=data as? Bool,value==false {
            throw Webview.Error("cannot delete properties of falsy (deleting \""+key+"\")");
        }
        if let bracketIndex=key.firstIndex(of:"]"),
           var array=data as? [Any?]{
            let symbol=key.prefix(upTo:bracketIndex);
            switch(symbol){
                case "pop","last":array.removeLast();break;
                case "shift":
                    if(!array.isEmpty){
                        array.remove(at:0);
                    }
                    break;
                case "push":array.append(nil);break;
                case "unshift":array.insert(nil,at:0);break;
                case "*":array.removeAll();break;
                default:
                    if let index=Int(symbol) {
                        array.remove(at:index);
                    };
                    break;
            }
            data=array;
        }
        else if var dictionary=data as? [String:Any] {
            dictionary.removeValue(forKey:key);
            data=dictionary;
        }
        else{
            throw ObjectCastError;
        }
    }
    
    func toObject()->[String:Any]{
        return store;
    }
}

let ObjectCastError=Webview.Error("cast to object failed");

File: src/browser/Store.js


module.exports={
    get:function(path){
        const store=JSON.parse(localStorage.getItem("store"))||{};
        const value=(()=>{
            if(path===undefined) return store;
            else if(typeof(path)==="string"){
                const keys=path.split(/\.|\[/g).filter(key=>key);
                let objects=[store];
                for(const key of keys){
                    objects=objects.flatMap(object=>{
                        const value=getProperty(key,object);
                        return key==="*]"?value:[value];
                    });
                }
                if(objects.length>1) return objects;
                else return objects[0];
            }
            else throw new Error("invalid path");
        })();
        if(path.includes("shift")||path.includes("pop")){
            localStorage.setItem("store",JSON.stringify(store));
        }
        return value;
    },
    set:function(path,value){
        let store;
        try{
            store=JSON.parse(localStorage.getItem("store"))||{};
            if(Array.isArray(path)){
                const {length}=path;
                if(length%2===0){
                    for(let i=0;i<length;i+=2){
                        const token=path[i],value=path[i+1];
                        mutate(store,token,value);
                    }
                }
                else throw new Error("array length should be even");
            }
            else{
                mutate(store,path,value);
            }
            localStorage.setItem("store",JSON.stringify(store));
            return store;
        }
        catch(error){
            localStorage.setItem("store",JSON.stringify(store));
            throw error;
        }
    },
};

const mutate=(data,path,value)=>{
    if(path&&typeof(path)==="string"){
        const keys=path.split(/\.|\[/g).filter(key=>key);
        const target=keys.pop();
        let objects=[data];
        if(keys.length){
            for(const key of keys){
                objects=objects.flatMap(object=>{
                    const value=getProperty(key,object);
                    return key==="*]"?value:[value];
                });
            }
        }
        for(const object of objects){
            if(value===undefined){
                deleteProperty(object,target);
            }
            else{
                setProperty(object,target,value);
            }
        }
    }
    else throw new Error("invalid key");
}

const getProperty=(key,data)=>{
    let value;
    const bracketi=key.indexOf("]");
    if(bracketi>-1){
        if(Array.isArray(data)){
            const index=key.substring(0,bracketi).trim();
            switch(index){
                case "*": value=data;break;
                case "last": value=data[data.length-1];break;
                case "pop": value=data.pop();break;
                case "shift": value=data.shift();break;
                default: 
                    const int=parseInt(index);
                    if(int>=0) value=data[int];
                    else throw UnrecognizedSymbol(index);
                break;
            }
        }
        else throw ArrayCastError(key);
    }
    else if(typeof(data)==="object"){
        value=data[key];
    }
    else throw ObjectCastError(key);
    return value;
}

const setProperty=(data,key,value)=>{
    const bracketi=key.indexOf("]");
    if(bracketi>-1){
        if(Array.isArray(data)){
            const index=key.substring(0,bracketi).trim();
            switch(index){
                case "*":
                    const {length}=data;
                    for(let i=0;i<length;i++){
                        data[i]=value;
                    }
                    break;
                case "last": data[data.length-1]=value;break;
                case "pop": data.pop();break;
                case "shift": data.shift();break;
                case "push": data.push(value);break;
                case "unshift": data.unshift(value);break;
                default: 
                    const int=parseInt(index);
                    if(int>=0) data[i]=value;
                    else throw UnrecognizedSymbol(index);
                break;
            }
        }
        else throw ArrayCastError(key);
    }
    else if(typeof(data)==="object"){
        data[key]=value;
    }
    else throw ObjectCastError(key);
}

const deleteProperty=(data,key)=>{
    const bracketi=key.indexOf("]");
    if(bracketi>-1){
        if(Array.isArray(data)){
            const index=key.substring(0,bracketi).trim();
            switch(index){
                case "last":
                case "pop": data.pop();break;
                case "shift": data.shift();break;
                case "push": data.push(null);break;
                case "unshift": data.unshift(null);break;
                case "*":data.splice(0,data.length);break;
                default:
                    const int=parseInt(index);
                    if(int>=0) data.splice(int,1);
                    else throw UnrecognizedSymbol(index);
                break;
            }
        }
        else throw ArrayCastError(key);
    }
    else if(typeof(data)==="object"){
        delete data[key];
    }
    else throw ObjectCastError(key);
}

const ArrayCastError=(key="")=>new Error(`cast to array failed at "${key}"`);
const ObjectCastError=(key="")=>new Error(`cast to object failed at "${key}"`);
const UnrecognizedSymbol=(symbol)=>new Error(`unrecognized symbol "${symbol}"`);
  